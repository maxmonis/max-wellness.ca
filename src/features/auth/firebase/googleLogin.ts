import { createDefaultSettings } from "@/features/settings/firebase/createDefaultSettings"
import { loadExerciseNames } from "@/features/settings/firebase/loadExerciseNames"
import { loadWorkoutNames } from "@/features/settings/firebase/loadWorkoutNames"
import { auth } from "@/firebase/app"
import { GoogleAuthProvider, signInWithPopup } from "firebase/auth"

/**
 * Opens a popup which prompts the user for their Google
 * credentials, then logs them in if they have an account
 * or creates a new account for them if not
 * @returns whether this is a new user
 */
export async function googleLogin() {
	const { user } = await signInWithPopup(auth, new GoogleAuthProvider())
	const [exerciseNames, workoutNames] = await Promise.all([
		loadExerciseNames(user.uid),
		loadWorkoutNames(user.uid),
	])
	const isNewUser =
		!Boolean(exerciseNames?.length) && !Boolean(workoutNames?.length)
	if (isNewUser) await createDefaultSettings(user.uid)
	return isNewUser
}
