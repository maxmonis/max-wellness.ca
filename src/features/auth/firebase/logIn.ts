import { auth } from "@/firebase/app"
import { signInWithEmailAndPassword } from "firebase/auth"

/**
 * Logs a user in using their email and password
 */
export function logIn(email: string, password: string) {
	return signInWithEmailAndPassword(auth, email, password)
}
