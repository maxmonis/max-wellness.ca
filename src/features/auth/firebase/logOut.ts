import { auth } from "@/firebase/app"
import { signOut } from "firebase/auth"

/**
 * Logs the user out
 */
export function logOut() {
	return signOut(auth)
}
