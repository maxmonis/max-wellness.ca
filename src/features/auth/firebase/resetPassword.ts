import { auth } from "@/firebase/app"
import { sendPasswordResetEmail } from "firebase/auth"

/**
 * Sends an email to the user allowing them to reset their password
 */
export function resetPassword(email: string) {
	return sendPasswordResetEmail(auth, email)
}
