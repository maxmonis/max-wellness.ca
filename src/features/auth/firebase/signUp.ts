import {
	defaultLiftNames,
	defaultWorkoutNames,
} from "@/features/settings/utils/constants"
import { auth, db } from "@/firebase/app"
import { createUserWithEmailAndPassword } from "firebase/auth"
import { addDoc, collection } from "firebase/firestore"

/**
 * Creates a new account using the provided name, email, and password
 */
export async function signUp(name: string, email: string, password: string) {
	const { user } = await createUserWithEmailAndPassword(auth, email, password)
	await addDoc(collection(db, "profile"), {
		liftNames: defaultLiftNames,
		photoURL: "",
		userId: user.uid,
		userName: name,
		workoutNames: defaultWorkoutNames,
	})
	return user
}
