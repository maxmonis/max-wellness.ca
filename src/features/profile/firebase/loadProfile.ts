import { db } from "@/firebase/app"
import { collection, getDocs, query, where } from "firebase/firestore"
import { isProfile } from "../utils/validators"

/**
 * @returns the user's profile from the database
 */
export async function loadProfile(userId: string) {
	const { docs } = await getDocs(
		query(collection(db, "profile"), where("userId", "==", userId)),
	)
	const profile = { ...docs[0]?.data(), id: docs[0]?.id }
	return isProfile(profile) ? profile : null
}
