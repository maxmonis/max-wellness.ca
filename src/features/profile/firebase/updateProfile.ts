import { db } from "@/firebase/app"
import { doc, updateDoc } from "firebase/firestore"
import { Profile } from "../utils/models"

/**
 * Updates the user's profile in the database
 */
export async function updateProfile(
	profileId: string,
	newFields: Partial<Profile>,
) {
	return updateDoc(doc(db, "profile", profileId), newFields)
}
