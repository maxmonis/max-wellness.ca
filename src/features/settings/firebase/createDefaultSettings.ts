import { db } from "@/firebase/app"
import { doc, writeBatch } from "firebase/firestore"
import { nanoid } from "nanoid"

export async function createDefaultSettings(userId: string) {
	const batch = writeBatch(db)
	for (const text of ["Bench Press", "Deadlift", "Squat"])
		batch.set(doc(db, "users", userId, "exerciseNames", nanoid()), { text })
	for (const text of ["Full Body", "Lower Body", "Upper Body"])
		batch.set(doc(db, "users", userId, "workoutNames", nanoid()), { text })
	await batch.commit()
}
