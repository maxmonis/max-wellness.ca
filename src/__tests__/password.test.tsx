import PasswordPage from "@/pages/password"
import { fireEvent, render, screen } from "@testing-library/react"
import * as firebaseAuth from "firebase/auth"
import { describe, expect, test, vi } from "vitest"

describe("password page", () => {
	test("renders initial state", () => {
		const { asFragment, unmount } = render(<PasswordPage />)
		screen.getByRole("heading", { name: "Reset Password" })
		screen.getByLabelText("Email")
		screen.getByRole("button", { name: "Send Link" })
		screen.getByRole("link", { name: "Return to Login" })
		screen.getByText("Need an account?")
		screen.getByRole("link", { name: "Sign Up" })
		expect(asFragment()).toMatchSnapshot()
		unmount()
	})

	test("validates input and handles submission", async () => {
		const sendPasswordResetEmailSpy = vi.fn()
		vi.spyOn(firebaseAuth, "sendPasswordResetEmail").mockImplementation(
			sendPasswordResetEmailSpy,
		)
		const { unmount } = render(<PasswordPage />)
		screen.getByRole("button", { name: "Send Link" }).click()
		await screen.findByText("Email is required")
		fireEvent.change(screen.getByLabelText("Email"), {
			target: { value: "not an email" },
		})
		expect(screen.queryByText("Email is required")).toBeNull()
		screen.getByRole("button", { name: "Send Link" }).click()
		await screen.findByText("Invalid email")
		fireEvent.change(screen.getByLabelText("Email"), {
			target: { value: "valid@email.test" },
		})
		expect(screen.queryByText("Invalid email")).toBeNull()
		screen.getByRole("button", { name: "Send Link" }).click()
		expect(sendPasswordResetEmailSpy).toBeCalledTimes(1)
		expect(sendPasswordResetEmailSpy).toBeCalledWith(
			undefined,
			"valid@email.test",
		)
		await screen.findByRole("heading", { name: "Email Sent" })
		screen.getByText(
			"Please check valid@email.test for instructions on resetting your password",
		)
		screen.getByRole("link", { name: "Return to Login" })
		unmount()
	})

	test("handles API errors", async () => {
		vi.spyOn(firebaseAuth, "sendPasswordResetEmail").mockImplementation(() => {
			throw Error("API password reset error")
		})
		const { unmount } = render(<PasswordPage />)
		fireEvent.change(screen.getByLabelText("Email"), {
			target: { value: "valid@email.test" },
		})
		screen.getByRole("button", { name: "Send Link" }).click()
		await screen.findByText("API password reset error")
		unmount()
	})
})
