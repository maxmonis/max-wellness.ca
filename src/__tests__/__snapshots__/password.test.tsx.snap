// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`password page > renders initial state 1`] = `
<DocumentFragment>
  <div
    class="dark fixed left-0 top-0 -z-10 bg-[url('/barbell.jpg')] bg-cover bg-[30%] bg-no-repeat object-cover"
  >
    <div
      class="relative h-screen w-screen overflow-auto px-4 pb-2 pt-10"
    >
      <div
        class="flex h-full flex-col items-center justify-between gap-20"
      >
        <div
          class="w-full max-w-xs rounded-lg border bg-background px-4 pb-6 pt-4 text-foreground sm:px-6"
        >
          <div>
            <div
              class="mb-4 flex w-full justify-center"
            >
              <span
                class="flex items-center gap-2 text-lg"
              >
                <img
                  alt="maxWellness logo"
                  class="h-6 w-6"
                  data-nimg="1"
                  decoding="async"
                  height="24"
                  loading="lazy"
                  src="/_next/image?url=%2Fandroid-chrome-192x192.png&w=48&q=75"
                  srcset="/_next/image?url=%2Fandroid-chrome-192x192.png&w=32&q=75 1x, /_next/image?url=%2Fandroid-chrome-192x192.png&w=48&q=75 2x"
                  style="color: transparent;"
                  width="24"
                />
                maxWellness
              </span>
            </div>
            <h1
              class="mb-2 font-bold"
            >
              Reset Password
            </h1>
          </div>
          <form
            class="grid gap-2"
            novalidate=""
          >
            <div
              class="relative"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                for="email"
              >
                Email
              </label>
              <input
                class="flex h-9 w-full rounded-md border bg-transparent px-3 py-1 text-sm text-inherit shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 border-input"
                id="email"
                maxlength="50"
                name="email"
                type="email"
                value=""
              />
            </div>
            <div
              class="mt-2 grid gap-4"
            >
              <button
                class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-10 px-4 py-2"
                type="submit"
              >
                Send Link
              </button>
            </div>
          </form>
          <footer
            class="mt-4 grid gap-2 text-sm"
          >
            <div>
              <a
                class="underline"
                href="/login"
              >
                Return to Login
              </a>
            </div>
            <div
              class="flex flex-wrap gap-x-1.5"
            >
              <p
                class="whitespace-nowrap"
              >
                Need an account?
              </p>
              <a
                class="underline"
                href="/register"
              >
                Sign Up
              </a>
            </div>
          </footer>
        </div>
        <a
          class="text-sm text-foreground"
          href="https://maxmonis.com"
          rel="noreferrer"
          target="_blank"
        >
          Â© Max Monis 2019-2024
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
